ext/drreg/drreg.c

```diff
@@ -503,6 +503,11 @@ drreg_insert_restore_all(void *drcontext, instrlist_t *bb, instr_t *inst,
             regs_restored[GPR_IDX(reg)] = false;
         if (!pt->reg[GPR_IDX(reg)].native) {
             if (force_restore || instr_reads_from_reg(inst, reg, DR_QUERY_INCLUDE_ALL) ||
+                /* Restore all app values prior to an annotation to ensure the
+                 * arguments passed to the handler are correct.
+                 */
+                (instr_is_label(inst) &&
+                 (ptr_uint_t)instr_get_note(inst) == DR_NOTE_ANNOTATION) ||
```

Comment:
Did you also plan to add a documentation note about this not working so well with multi-phase use?

Reply:
I'm actually a little confused about that: if this annotation-triggered restore of app values won't work well with multi-phase, then wouldn't an app instruction reading a register have the same problem (this is triggering that same code at the same points)?  And so would a client inserting a clean call, but there's no mention of multi-phase in drreg.dox sec_drreg_app_values.

Reply:
> then wouldn't an app instruction reading a register have the same problem

Yes, if a user reserves a gpr in app2app phase, then the responsibility to restore the gpr before an app instruction that reads it (and also to remember that an app write will clobber the meta value in that reg) lies with the user.

>  And so would a client inserting a clean call,

You mean a client that sends current reg values as args to a clean call? Yes, the same issue.

I think I should document these caveats of out-of-insertion-phase drreg use better. Assigned #5121 to myself.

Reply:
> > then wouldn't an app instruction reading a register have the same problem
> 
> Yes, if a user reserves a gpr in app2app phase, then the responsibility to restore the gpr before an app instruction that reads it (and also to remember that an app write will clobber the meta value in that reg) lies with the user.

But that's not enough to restore in the app2app phase as it will mess up the use of the reg in the insertion phase: the insertion phase can easily have the wrong value in its slot (it will have the tool value) and it will restore that most likely after the explicit app2app restore, resulting in the wrong value read by the app and subsequent corruption.

Reply:
I put that into #5121.  But the fundamental issues seem to have nothing to do with annotations other than annotations adding an additional app read (but that's a small fraction of the real app reads).  Since I'm already making other minor changes here I'll add a #5121 comment for the general app read restore.